src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                // Nuget should not wait for projects that failed DTB
+                if (!_enabled || SourceBlock.Completion.IsFaulted || SourceBlock.Completion.IsCompleted)
+                {
+                    return false;
+                }
+
+                // Avoid possible deadlock.
+                // Because RestoreCoreAsync() is called inside a dataflow block it will not be called with new data
+                // until the old task finishes. So, if the project gets nominating restore, it will not get updated data.
+                if (IsNuGetRestoreOnGoing())
+                {
+                    return false;
+                }
+
+                ConfiguredProject? activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                // After the first nomination, we should check the saved nominated version
+                return IsSavedNominationEmptyOrOlder(activeConfiguredProject);
+            }
+        }
+
+        private bool IsNuGetRestoreOnGoing()
+        {
+            // If NominateForRestoreAsync() has not finished, return false
+            return _nugetRestoreStarted;
+        }
+
+        protected virtual bool IsSavedNominationEmptyOrOlder(ConfiguredProject activeConfiguredProject)
+        {
+            if (!_savedNominatedConfiguredVersion.TryGetValue(activeConfiguredProject.ProjectConfiguration,
+                    out IComparable latestSavedVersionForActiveConfiguredProject) ||
+                activeConfiguredProject.ProjectVersion.IsLaterThan(latestSavedVersionForActiveConfiguredProject))
+            {
+                return true;
+            }
+
+            if (_savedNominatedConfiguredVersion.Count == 1)
+            {
+                return true;
+            }
+
+            foreach (var loadedProject in activeConfiguredProject.UnconfiguredProject.LoadedConfiguredProjects)
+            {
+                if (_savedNominatedConfiguredVersion.TryGetValue(loadedProject.ProjectConfiguration, out IComparable savedProjectVersion))
+                {
+                    if (loadedProject.ProjectVersion.IsLaterThan(savedProjectVersion))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        protected virtual bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput>? configuredInputs)
+        {
+            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
```

Comment:
I don't get why this method needs the double-walk @lifengl? Why does the first walk need to occur?

Reply:
it is largely duplicated from IsSavedNominationEmptyOrOlder, so i didn't pay too much attention. The first check is to ensure active configured project is included (instead of processing old data during configuration switching), and for majority of projects with single target framework, it won't need run into the slow double loop code.